flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -540,15 +540,12 @@ private boolean performCheckpoint(CheckpointMetaData checkpointMetaData) throws
 
 		synchronized (lock) {
 			if (isRunning) {
+				checkpointState(checkpointMetaData);
 
-				// Since both state checkpointing and downstream barrier emission occurs in this
-				// lock scope, they are an atomic operation regardless of the order in which they occur.
-				// Given this, we immediately emit the checkpoint barriers, so the downstream operators
-				// can start their checkpoint work as soon as possible
+				// broadcast barriers after snapshot operators' states.
 				operatorChain.broadcastCheckpointBarrier(
-						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp());
-
-				checkpointState(checkpointMetaData);
+						checkpointMetaData.getCheckpointId(), checkpointMetaData.getTimestamp()
+				);
```

Comment:
I think the important performance characteristic is not how the locks perform when there is only a single thread trying to acquire the lock but if there is some kind of lock contention by multiple threads. If you change your benchmark so that you have multiple threads running at the same time, you will see that the `ReentrantLock` and the `ReentrantReadWriteLock` are much more expensive than the `Object`.